### PR TITLE
[docs] update attention implementation and cache docs

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

As per title, and we will delete `Legacy cache format` section in 2-3 releases when all legacy support is removed. 

This PR:
- Adds docs on multimodal attention implementation setting
- Cross references attn implementation docs in other pages. I didn't know that docs page existed and searching for FA2/SDPA usually doesn't return it. Let's make it more discoverable
- Adds a section on what is cache position, as reported by many users concept of `cache position` is still confusing. We can add later more examples, I remember @gante had a PR on fixing generate when cache position is provided by users :)

### CODE REVIEW

File: docs/source/en/attention_interface.md [hunk 1]
```diff
@@ -72,6 +72,34 @@ model(torch.ones(1, 5, dtype=int))
 and it will stop printing the statements, as it now uses the `sdpa` attention.  
 This allows to quickly change an attention function, without needing to reload the model!
 
+## Different attention per backbone in multimodal models
+
+For multimodal models, you may want to load different backbones with different attention functions. For example, some vision backbones perform better with full precision only and are incompatible with Flash Attention. If you want to take advantage of Flash Attention while keeping your vision encoder in fp32, you can configure different attention implementations per backbone.
```
Comment (stevhliu) [inline docs/source/en/attention_interface.md]:
```suggestion
Attention functions may work better for different backbones. For example, some vision backbones perform better in fp32, but are incompatible with FlashAttention. To continue using FlashAttention while keeping the vision encoder in fp32, create a dict and map each config to an attention implementation as shown below.
```

File: docs/source/en/cache_explanation.md [hunk 2]
```diff
@@ -132,6 +132,35 @@ for _ in range(max_new_tokens):
 print(tokenizer.batch_decode(generated_ids, skip_special_tokens=True)[0])
 "[INST] Hello, what's your name. [/INST]  Hello! My name is LLaMA,"
 ```
+
+### Cache Position Usage
```
Comment (stevhliu) [inline docs/source/en/cache_explanation.md]:
```suggestion
## Cache position
```

File: docs/source/en/cache_explanation.md [hunk 3]
```diff
@@ -132,6 +132,35 @@ for _ in range(max_new_tokens):
...
+
+The cache position tracks where the new tokens should be inserted in the attention cache. It represents the *absolute* position of each token in the context, independent of padding or batch structure. Suppose that you have already cached `N` tokens and you are now processing `K` new tokens. Then the cache position for the new tokens will range from `N` to `N + K - 1`. In other words, you're processing tokens at positions - `[N, N + 1, N + 2, ..., N + K - 1]`
```
Comment (stevhliu) [inline docs/source/en/cache_explanation.md]:
```suggestion
The cache position tracks where to insert new tokens in the attention cache. It represents the *absolute* position of each token in the context, independent of padding or batch structure. Suppose you already cached `N` tokens and are now processing `K` new tokens. The cache position for the new tokens will range from `N` to `N + K - 1`. In other words, you're processing tokens at positions - `[N, N + 1, N + 2, ..., N + K - 1]`.
```

File: docs/source/en/cache_explanation.md [hunk 4]
```diff
@@ -132,6 +132,35 @@ for _ in range(max_new_tokens):
...
+
+Cache position is used internally for two purposes:
+
+1. Selecting new tokens to process in the input sequence and ensuring only tokens that haven’t been cached yet are passed to the model's `forward`.
+2. Storing key/value pairs at the correct positions in the cache. Especially important for fixed-size caches like [`StaticCache`] that preallocate certain length for cache in advance.
```
Comment (stevhliu) [inline docs/source/en/cache_explanation.md]:
```suggestion
2. Storing key/value pairs at the correct positions in the cache. This is especially important for fixed-size caches, like [`StaticCache`], that pre-allocates a specific cache length.
```

File: docs/source/en/cache_explanation.md [hunk 5]
```diff
@@ -132,6 +132,35 @@ for _ in range(max_new_tokens):
...
+
+Usually you don't have to worry about cache position as the generation loop takes care of it. Yet, in cases where you are writing your own custom generation method, it is important that cache positions are accurate since they are used to write and read key/value states into fixed slots.
```
Comment (stevhliu) [inline docs/source/en/cache_explanation.md]:
```suggestion
The generation loop usually takes care of the cache position, but if you're writing a custom generation method, it is important that cache positions are accurate since they are used to write and read key/value states into fixed slots.
```

File: docs/source/en/cache_explanation.md [hunk 6]
```diff
@@ -132,6 +132,35 @@ for _ in range(max_new_tokens):
...
+
+There is an example usage with simple custom generation loop above, so let's create a more complex example here.
```
Comment (stevhliu) [inline docs/source/en/cache_explanation.md]:
```suggestion
```

File: docs/source/en/llm_optims.md [hunk 7]
```diff
@@ -341,7 +341,7 @@ A known issue with transformer models is that the self-attention mechanism grows
 
 FlashAttention and [FlashAttention-2](./perf_infer_gpu_one#flashattention-2) break up the attention computation into smaller chunks and reduces the number of intermediate read/write operations to the GPU memory to speed up inference. FlashAttention-2 improves on the original FlashAttention algorithm by also parallelizing over sequence length dimension and better partitioning work on the hardware to reduce synchronization and communication overhead.
 
-To use FlashAttention-2, set [attn_implementation](https://hf.co/docs/transformers/main/en/main_classes/text_generation#transformers.PreTrainedModel.from_pretrained.attn_implementation) to `"flash_attention_2"` in [`~PreTrainedModel.from_pretrained`].
+To use FlashAttention-2, set [attn_implementation](https://hf.co/docs/transformers/main/en/main_classes/text_generation#transformers.PreTrainedModel.from_pretrained.attn_implementation) to `"flash_attention_2"` in [`~PreTrainedModel.from_pretrained`] or set with `model.set_attention_implementation("flash_attention_2")` to dynamically update the [attention interface](./attention_interface) after the model is loaded.
```
Comment (stevhliu) [inline docs/source/en/llm_optims.md:344]: May be helpful to add `model.set_attention_implementation("flash_attention_2")` to the code snippet

File: docs/source/en/perf_infer_gpu_one.md [hunk 8]
```diff
@@ -175,7 +175,7 @@ There are three supported implementations available.
 - [xFormers](https://github.com/facebookresearch/xformers) or Memory-Efficient Attention is able to support models with the fp32 torch type.
 - C++ implementation of scaled dot product attention
 
-SDPA is used by default for PyTorch v2.1.1. and greater when an implementation is available. You could explicitly enable SDPA by setting `attn_implementation="sdpa"` in [`~PreTrainedModel.from_pretrained`] though. Certain attention parameters, such as `head_mask` and `output_attentions=True`, are unsupported and returns a warning that Transformers will fall back to the (slower) eager implementation.
+SDPA is used by default for PyTorch v2.1.1. and greater when an implementation is available. You could explicitly enable SDPA by setting `attn_implementation="sdpa"` in [`~PreTrainedModel.from_pretrained`] though. Certain attention parameters, such as `head_mask` and `output_attentions=True`, are unsupported and returns a warning that Transformers will fall back to the (slower) eager implementation. You can also change the attention implementation after loading the model for most architectures. For details, check out the [attention interfaces documentation](./attention_interface).
```
Comment (stevhliu) [inline docs/source/en/perf_infer_gpu_one.md]:
```suggestion
SDPA is used by default for PyTorch v2.1.1. and greater when an implementation is available. You could explicitly enable SDPA by setting `attn_implementation="sdpa"` in [`~PreTrainedModel.from_pretrained`] though. Certain attention parameters, such as `head_mask` and `output_attentions=True`, are unsupported and returns a warning that Transformers will fall back to the (slower) eager implementation.

Refer to the [AttentionInterface](./attention_interface) guide to learn how to change the attention implementation after loading a model.
```

File: docs/source/en/perf_infer_gpu_one.md [hunk 9]
```diff
@@ -234,7 +234,7 @@ FlashAttention2 support is currently limited to Instinct MI210, Instinct MI250 a
 </hfoption>
 </hfoptions>
 
-Enable FlashAttention2 by setting `attn_implementation="flash_attention_2"` in [`~PreTrainedModel.from_pretrained`]. FlashAttention2 is only supported for models with the fp16 or bf16 torch type. Make sure to cast your model to the appropriate data type first.
+Enable FlashAttention2 by setting `attn_implementation="flash_attention_2"` in [`~PreTrainedModel.from_pretrained`] or by setting `model.set_attention_implementation("flash_attention_2")` to dynamically update the [attention interface](./attention_interface). FlashAttention2 is only supported for models with the fp16 or bf16 torch type. Make sure to cast your model to the appropriate data type first.
```
Comment (stevhliu) [inline docs/source/en/perf_infer_gpu_one.md:243]: Also add `model.set_attention_implementation` snippet here

**Review by stevhliu — APPROVED** [review]
Thanks for the updates and improved cross-linking!
